llm.py

```diff
@@ -68,16 +69,41 @@ def gen(
                 output_hidden_states=return_hiddens,
                 return_dict_in_generate=True,
                 use_cache=True,
-                **args
+                **modelArgs
             )
             ts = generate_dict.sequences
 
             def helper(tid):
-                return tid not in tokenizer.all_special_ids
-
-            ntokens = sum(sum(helper(tid) for tid in t) for t in ts)
+                return tid in tokenizer.all_special_ids
+            # This code needs uses the token count in two different contexts:
+            # 1. for measuring the amount of generated tokens (as a proxy for time)
+            # 2. (as a workaround for StarCoder) for knowing where the newly-generated
+            # response begins (the model always returns both prompt and response).
+            #
+            # For (1), we ideally do not want (too many) special tokens, especially
+            # no filler tokens.
+            # For (2), we need to count all tokens.
+            #
+            # As a compromise, this code assumes that there will be barely any special
+            # tokens, and checks this below.
+            specialtok = sum(sum(helper(tid) for tid in t) for t in ts)
+            assert(specialtok < 2) # just to confirm there is no need to count special tok
```

Comment:
why is 2 a magic number?

Reply:
I have never seen more than two special tokens per run (across all invocations). I am guessing the two are something like start and stop tokens?